### DISCUSSION TITLE
Logical Switch "Sticky" Behaviour

### DISCUSSION THREAD

**Discussion by EhAye** [EdgeTX/edgetx#2587] [Q&A] [answered]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Transmitter firmware

### Current Behavior

Flicking SH Down then back Up, the logical function doesn't know SH is back Up. 
So when i flip SF Down, the sticky activates, as if SH was still up.

Toggling SF up and down won't activate the sticky (correct behaviour) but if i switch SH Down and back Up, SF will toggle the sticky upon next Down.

For clarity: the purpose of this function is a dual-switch arming system, with 1 switch to disarm.

L01:
Func: Sticky
V1: SH Down
V2: SF Up
AND Switch: SF Down
Duration: 0
Delay: 0

### Expected Behavior

The sticky should only be activated when both SF and SH are Down at the SAME TIME.

when i hold SH Down AND SF Down, the sticky should activate
When i release SH to Up, the sticky should remain active.
When i switch SF to Up, the sticky should Deactivate.

### Steps To Reproduce

Toggle HS Up
Toggle HS Down
Toggle FS Up

Sticky activates even though HS was down. 

### Version

2.7.1

### Transmitter

Radiomaster TX16S / TX16SMK2

### Anything else?

both 2.7.1 and 2.8 RC3... Same behaviour.

Reversing the switches in the function produces a different wrong behaviour:

L01:
Func: Sticky
V1: SF Down
V2: SF Up
AND Switch: SH Down
Duration: 0
Delay: 0

This no longer "sticks". It behaves like an AND and not a Sticky
I would expect this to activate when SF and SH are down, and remain active until SF is up, regardless of SH.

These Sticky function is sticky where it shouldn't be, and non-sticky where it should be!

This work-around behaves correctly, but requires 2 functions: 
L02 activates when both SF and SH switches are Down,
L01 activates when L02 becomes active
L01 only deactivates when SF goes Up.

L01:
Func: Sticky
V1: L02
V2: SF Up
AND Switch: ---
Duration: 0
Delay: 0

L02:
Func: AND
V1: SH Down
V2: SF Down
AND Switch ---
Duration: 0.1
Delay: 0

**Comment by pfeerick** [comment, marked as answer]
I beleive this to be intended behaviour (it is certainly hereditary as OpenTX Sticky logic switch behaves exactly the same) and thus a  misunderstanding of the Sticky / Set Reset latch behaviour, not a bug. By pulling SH down, you have met the Set condition. _However_, I would also have have thought that the AND switch condition needed to be valid at the same time, this then presents a problem... does the AND switch need to be valid for set AND reset? Which really should be yes. Thus, in your scenario, you will never get it to reset, as SF would need to be in the down position for the reset condition to be processed (which is SFup). 😆 

To see the sticky switch logic properly, you really need to use two different switches for V1 and V2 if you're going to AND it with another switch... i.e. set V1 to SF down, and V2 to SA down, and AND to SH down. You'll then see that as long as you're holding SH down, when you move SF from up to down, the LS will set/latch. as long as you're holding SH down, the logical switch will stay active, regardless of the position of SF. But, when you move SA down, the LS will be reset. And if you release SH at any time, the logical switch will also be negated. But, it still remembers it's set/reset state, so if you release SH when the logical switch was set, it will become active again when you press SH down again.

To achieve what you are after, a two stage LS is the way to do it. The following will trigger L01 when both SH and SF are down. L02 then looks for that, stays set if both switches are down, and gets reset if SF then goes up. So L02 becomes your arming switch. 

L01:
Func: AND
V1: SF down
V2: SH down
AND Switch: ---
Duration: 0
Delay: 0

L02:
Func: Stcky
V1: L01
V2: SF up
AND Switch: ---
Duration: 0
Delay: 0

**Reply by EhAye** [reply]
I guess it would cause far more problems to change it now, if this is what people's configurations are expecting. I thought for sure this was a bug, but i can see how its working now with your description.

I would have liked the AND field to only apply for V1 and not V2 :)

**Reply by TomAlperin** [reply]
When I set up a logical arming switch as you describe with two stages, it is still possible to get stuck armed if the switches are flipped quickly. There is a race condition where the logical armed of the first switch has not happened before the disarm (second logical switch). Since it's tracking the switch change, it does not recognize the disarm has happened unless the switch is flipped back and fourth a second time. This makes sticky switches behave unreliably.

**Comment by TomAlperin** [comment]
Below is the normal operation of the sticky switch. The L3 switch is just there to show the state of the armed switch.

![sticky-normal-operation](https://github.com/EdgeTX/edgetx/assets/6037295/646e25be-e634-4ca4-a9c5-f7316e0e931f)

Here is the error state. Notice that the armed switch is off but the sticky is still on. This is dangerous if used for arming and I wouldn't trust sticky switches for anything after discovering this. This happens on OpenTX as well but that does not mean it is not a bug. This occurs if the switch is flipped back quickly.

![sticky-error-condition](https://github.com/EdgeTX/edgetx/assets/6037295/fa6865dd-13e6-484e-a4d9-7b148202bc76)

**Comment by philmoz** [comment]
> Here is the error state. Notice that the armed switch is off but the sticky is still on. This is dangerous if used for arming and I wouldn't trust sticky switches for anything after discovering this. This happens on OpenTX as well but that does not mean it is not a bug. This occurs if the switch is flipped back quickly.

Set the AND property of L02 to L01.

**Reply by TomAlperin** [reply]
This is not an acceptable solution. I already experimented with possible work-around solutions including setting the AND property but the results were still unpredictable. If you set the AND it creates a situation where disarm appears to work but pre-arm is not required the next time you arm. The sticky switch needs to somehow check the state more often or there needs to be a guarantee that the disarm condition has a short delay to guarantee the check does not happen before the sticky switch triggers. At the very least, users should be made aware that using sticky switches this way is unreliable and should not be used for pre arm.

**Comment by philmoz** [comment]
> This is not an acceptable solution. I already experimented with possible work-around solutions including setting the AND property but the results were still unpredictable. 

I tested your example on my TX16S with L02 AND set to L01 and it behaves exactly as you would expect it to.
L02 will only be set when L01 changes from off to on.

> If you set the AND it creates a situation where disarm appears to work but pre-arm is not required the next time you arm.

Can you explain what you are expecting to happen here?
With L02 AND = L01, you can only activate L02 by changing ARM to up with PRE down or changing PRE to down with ARM up.

**Reply by TomAlperin** [reply]
What I expect to happen is the state illustrated in the second image never occurs no matter how fast you flip the arm switch on and off. When I added the AND condition thinking this might fix it. Disarm worked but arm also worked after but without the pre arm switch. Adding the AND condition shouldn't be necessary if the sticky switch behaved predictably and all the guides on YouTube and other sites do not include the AND condition. The way I think this should work is that L2 should always turn off from the second condition no matter how fast you flip the switch.

**Comment by philmoz** [comment]
> Disarm worked but arm also worked after but without the pre arm switch.

This does not happen on my TX16S - what radio are you using?

**Reply by TomAlperin** [reply]
I have a TX16s Max. I can try setting it up again to verify but pretty sure pre arm was not required after a fast arm switch toggle.

I just tried setting L1 on the AND of the sticky switch. What this does is make the arm not stick when the pre arm is released. This is not what I had tried previously. The other thing that I tried is to set the arm switch on position to the AND this appears to work but after a fast toggle it will arm without using the pre arm.
